In [1]:
import json
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [2]:
# Load in the data
metaPath = 'Data/meta_Magazine_Subscriptions.json'
reviewPath = 'Data/Magazine_Subscriptions.json'

metaData = []
reviewData = []

for line in open(metaPath, 'r'):
    metaData.append(json.loads(line))
    
for line in open(reviewPath, 'r'):
    reviewData.append(json.loads(line))

In [3]:
# convert metadata into pandas dataframe

metaDF = pd.DataFrame.from_dict(metaData)
metaDF.head()

category tech1  \
0  [Magazine Subscriptions, Professional & Educat...         
1  [Magazine Subscriptions, Arts, Music &amp; Pho...         
2  [Magazine Subscriptions, Fashion &amp; Style, ...         
3  [Magazine Subscriptions, Sports, Recreation & ...         
4  [Magazine Subscriptions, Professional & Educat...         

                                         description fit  \
0  [REASON is edited for people interested in eco...       
1  [Written by and for musicians. Covers a variet...       
2  [Allure is the beauty expert. Every issue is f...       
3  [FLIGHT JOURNAL includes articles on aviation ...       
4  [RIDER is published for the road and street ri...       

                                           title  \
0  <span class="a-size-medium a-color-secondary"   
1  <span class="a-size-medium a-color-secondary"   
2  <span class="a-size-medium a-color-secondary"   
3  <span class="a-size-medium a-color-secondary"   
4  <span class="a-size-medium a-color-secondary"   

                                            also_buy tech2  \
0  [B002PXVYLE, B01MCU84LB, B000UHI2LW, B01AKS14A...         
1  [B002PXVYGE, B0054LRNC8, B000BVEELE, B00006KC3...         
2  [B001THPA4O, B002PXVZWW, B001THPA1M, B001THPA1...         
3  [B07JVF7QW4, B00ATQ6FPY, B002G551F6, B00008CGW...         
4  [B002PXVYD2, B01BM7TOU6, B000060MKJ, B000BNNIG...         

                               brand feature                             rank  \
0                    Reason Magazine      []                               []   
1           String Letter Publishers      []  742 in Magazine Subscriptions (   
2            Conde Nast Publications      []                               []   
3                  AirAge Publishing      []                               []   
4  EPG Media & Specialty Information      []                               []   

                                           also_view  \
0  [B002PXVYLE, B000UHI2LW, B01MCU84LB, B002PXW18...   
1  [B002PXVYGE, B0054LRNC8, B00006L16A, 171906487...   
2  [B002PXVZWW, B001THPA4O, B001THPA1M, B01N819UD...   
3  [B002G551F6, B00ATQ6FPY, B00005N7PT, B001THPA2...   
4  [B01BM7TOU6, B000060MKJ, B002PXVYD2, B000BNNIG...   

                                             details                main_cat  \
0  {'Format:': 'Print Magazine', 'Shipping: ': 'C...  Magazine Subscriptions   
1  {'Format:': 'Print Magazine', 'Shipping: ': 'C...  Magazine Subscriptions   
2  {'Format:': 'Print Magazine', 'Shipping: ': 'C...  Magazine Subscriptions   
3  {'Format:': 'Print Magazine', 'Shipping: ': 'C...  Magazine Subscriptions   
4  {'Format:': 'Print Magazine', 'Shipping: ': 'C...  Magazine Subscriptions   

  similar_item date price        asin  \
0                          B00005N7NQ   
1                          B00005N7OC   
2                          B00005N7OD   
3                          B00005N7O9   
4                          B00005N7O6   

                                            imageURL  \
0  [https://images-na.ssl-images-amazon.com/image...   
1  [https://images-na.ssl-images-amazon.com/image...   
2  [https://images-na.ssl-images-amazon.com/image...   
3  [https://images-na.ssl-images-amazon.com/image...   
4  [https://images-na.ssl-images-amazon.com/image...   

                                     imageURLHighRes  
0  [https://images-na.ssl-images-amazon.com/image...  
1  [https://images-na.ssl-images-amazon.com/image...  
2  [https://images-na.ssl-images-amazon.com/image...  
3  [https://images-na.ssl-images-amazon.com/image...  
4  [https://images-na.ssl-images-amazon.com/image...

In [4]:
# convert reviews data into pandas dataframe

reviewDF = pd.DataFrame.from_dict(reviewData)
reviewDF.head()

overall vote  verified   reviewTime      reviewerID        asin  \
0      5.0    9     False   11 8, 2001   AH2IFH762VY5U  B00005N7P0   
1      5.0    9     False  10 31, 2001   AOSFI0JEYU4XM  B00005N7P0   
2      3.0   14     False  03 24, 2007  A3JPFWKS83R49V  B00005N7OJ   
3      5.0   13     False  11 10, 2006  A19FKU6JZQ2ECJ  B00005N7OJ   
4      5.0  NaN      True  07 14, 2014  A25MDGOMZ2GALN  B00005N7P0   

         reviewerName                                         reviewText  \
0        ted sedlmayr  for computer enthusiast, MaxPC is a welcome si...   
1     Amazon Customer  Thank god this is not a Ziff Davis publication...   
2         Bryan Carey  Antiques Magazine is a publication made for an...   
3  Patricia L. Porada  This beautiful magazine is in itself a work of...   
4               Alvey                          A great read every issue.   

                                           summary  unixReviewTime  \
0           AVID READER SINCE "boot"  WAS THE NAME      1005177600   
1                               The straight scoop      1004486400   
2  Antiques Magazine is Good, but not for Everyone      1174694400   
3                           THE  DISCERNING READER      1163116800   
4                                       Five Stars      1405296000   

                            style image  
0                             NaN   NaN  
1                             NaN   NaN  
2  {'Format:': ' Print Magazine'}   NaN  
3  {'Format:': ' Print Magazine'}   NaN  
4                             NaN   NaN

In [5]:
# Drop duplicates

#convert dataframe values to string in order to use drop_duplicates method
metaDF = metaDF.astype(str)
reviewDF = reviewDF.astype(str)

metaDF = metaDF.drop_duplicates()
reviewDF = reviewDF.drop_duplicates()

In [6]:
# Look at the columns in metaDF
metaDF.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2320 entries, 0 to 3384
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   category         2320 non-null   object
 1   tech1            2320 non-null   object
 2   description      2320 non-null   object
 3   fit              2320 non-null   object
 4   title            2320 non-null   object
 5   also_buy         2320 non-null   object
 6   tech2            2320 non-null   object
 7   brand            2320 non-null   object
 8   feature          2320 non-null   object
 9   rank             2320 non-null   object
 10  also_view        2320 non-null   object
 11  details          2320 non-null   object
 12  main_cat         2320 non-null   object
 13  similar_item     2320 non-null   object
 14  date             2320 non-null   object
 15  price            2320 non-null   object
 16  asin             2320 non-null   object
 17  imageURL         2320 non-null   

In [7]:
# Look at sample cell values in metaDF 
pd.set_option('display.max_colwidth', None)
metaDF.iloc[[0]]

category  \
0  ['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'Economics & Economic Theory']   

  tech1  \
0         

                                                                                                                                                                                                                                                                                                                                                                                             description  \
0  ['REASON is edited for people interested in economic, social, and international issues. Viewpoint stresses individual liberty, private responsibility, and limited government. Some emphasis on Pacific Rim, local/state issues with national impact, science/technology. Regular departments include news/trends, book reviews (mostly history, politics, and economics), and cultural commentary.']   

  fit                                          title  \
0      <span class="a-size-medium a-color-secondary"   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           also_buy  \
0  ['B002PXVYLE', 'B01MCU84LB', 'B000UHI2LW', 'B01AKS14AQ', 'B002PXVYPU', 'B001THPA26', '1476752842', 'B002CT515Q', 'B00KQ0HP2K', 'B00XZF1JUM', 'B0058EONOM', 'B01FV51RKA', 'B0032KHQTS', 'B002GCU2S0', 'B079JCLNZ4', 'B002PXW18E', 'B00005NIOH', 'B00005N7SD', 'B002LDA9VY', '0345816021', 'B00006KX3K', 'B002GCU2SA', 'B0047VIALE', 'B00007AZRH', 'B06XWB4X41', '0465058728', 'B0045VPFDW', 'B00FP59VGY', 'B001THPA3K', 'B002BFZ9NA', '0735224897', 'B01F2MKW0I', 'B0036QY66Y', '0415922224', '0226315398', 'B00005N7R5', 'B002KUNVD6', 'B00005N7PT', 'B0025ZOVEO', 'B002PXW0M6', 'B002PXVYEQ', 'B002PXVYGO', 'B002PXVYVE', '1548650218', '0465028616', 'B01FIR6AJ4', '1541645375', 'B003I7HS3Y', '0061253804', '1491904224', '1944424768', 'B001TP73UU', 'B00EV5HI0K', '0062308270', '0525540075', 'B06XWC6PRS', 'B002PXW0LW', 'B002PXVYBE', 'B001THPA58', 'B004GVZUUY', 'B001TP73UA', 'B000063XJR', 'B002PXVYO6', '0465060730', '1594204225', 'B002PXW24C', 'B002XPVNOC', 'B002CT51A6', '0998289027', '0061479012', '0525427570', 'B001U5SPLK']   

  tech2            brand feature rank  \
0        Reason Magazine      []   []   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            also_view  \
0  ['B002PXVYLE', 'B000UHI2LW', 'B01MCU84LB', 'B002PXW18E', 'B00KQ0HP2K', 'B002GCU2S0', 'B002GCU2SA', 'B01AKS14AQ', 'B00

From the metadata, I will extract the following features:
- category
- description
- brand

In [8]:
# Exploring the types of values in the feature
metaDF[['category']].value_counts()

category                                                                                                                                           
[]                                                                                                                                                     442
['Magazine Subscriptions', 'Lifestyle & Cultures']                                                                                                      64
['Magazine Subscriptions', 'Fashion & Style', 'Women']                                                                                                  43
['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Transportation']                                             42
['Magazine Subscriptions', 'Travel, City & Regional', 'United States', 'South']                                                                         39
                                                                             

In [9]:
metaDF[['description']].value_counts()

description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
[]                                                                                                                                                                                                                                                                                                                                                                                                                                                                           

We see that there are missing values filled in as '[]' in both the category and description column.  This has to be cleaned up.

In [10]:
metaDF[['brand']].value_counts()

brand                   
Hearst Magazines            255
Conde Nast                  184
                            147
Meredith                     48
Bonnier Corporation          43
                           ... 
Imagine That II, LLC          1
Imb Marketing Ltd Unit 6      1
Imbibe Media Inc              1
Below Zero Media              1
Zyzzyva /Laura Cogan          1
Length: 948, dtype: int64

In [11]:
# Replace rows with values '[]' to an empty string
features = ['category', 'description', 'brand']

for feature in features:
    metaDF[feature] = metaDF[feature].apply(lambda x: '' if x == '[]' else x)

In [12]:
pd.set_option('display.max_colwidth', 20)

# Combining relevant features into a single feature

def combined_features(row):
    combined_features = ''
    for feature in features:
        combined_features = combined_features + ' ' + row[feature] 
    return combined_features

metaDF['combined_features'] = metaDF.apply(combined_features, axis = 1)
metaDF.head()

category tech1          description fit                title  \
0  ['Magazine Subsc...        ['REASON is edit...      <span class="a-s...   
1  ['Magazine Subsc...        ['Written by and...      <span class="a-s...   
2  ['Magazine Subsc...        ['Allure is the ...      <span class="a-s...   
3  ['Magazine Subsc...        ['FLIGHT JOURNAL...      <span class="a-s...   
4  ['Magazine Subsc...        ['RIDER is publi...      <span class="a-s...   

              also_buy tech2                brand feature  \
0  ['B002PXVYLE', '...            Reason Magazine      []   
1  ['B002PXVYGE', '...        String Letter Pu...      []   
2  ['B001THPA4O', '...        Conde Nast Publi...      []   
3  ['B07JVF7QW4', '...          AirAge Publishing      []   
4  ['B002PXVYD2', '...        EPG Media & Spec...      []   

                  rank            also_view              details  \
0                   []  ['B002PXVYLE', '...  {'Format:': 'Pri...   
1  742 in Magazine ...  ['B002PXVYGE', '...  {'Format:': 'Pri...   
2                   []  ['B002PXVZWW', '...  {'Format:': 'Pri...   
3                   []  ['B002G551F6', '...  {'Format:': 'Pri...   
4                   []  ['B01BM7TOU6', '...  {'Format:': 'Pri...   

              main_cat similar_item date price        asin  \
0  Magazine Subscri...                          B00005N7NQ   
1  Magazine Subscri...                          B00005N7OC   
2  Magazine Subscri...                          B00005N7OD   
3  Magazine Subscri...                          B00005N7O9   
4  Magazine Subscri...                          B00005N7O6   

              imageURL      imageURLHighRes    combined_features  
0  ['https://images...  ['https://images...   ['Magazine Subs...  
1  ['https://images...  ['https://images...   ['Magazine Subs...  
2  ['https://images...  ['https://images...   ['Magazine Subs...  
3  ['https://images...  ['https://images...   ['Magazine Subs...  
4  ['https://images...  ['https://images...   ['Magazine Subs...

In [13]:
# Extracting features 
cv = CountVectorizer()
count_matrix = cv.fit_transform(metaDF['combined_features'])
print("Count Matrix:", count_matrix.toarray())

Count Matrix: [[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [14]:
#cosine similarity matrix
cosine_sim = cosine_similarity(count_matrix)
pd.DataFrame(cosine_sim)

0         1         2         3         4         5         6     \
0     1.000000  0.283333  0.264851  0.223146  0.389213  0.287085  0.279508   
1     0.283333  1.000000  0.381385  0.289076  0.358253  0.334932  0.307459   
2     0.264851  0.381385  1.000000  0.270787  0.362666  0.480538  0.408635   
3     0.223146  0.289076  0.270787  1.000000  0.290701  0.256247  0.331702   
4     0.389213  0.358253  0.362666  0.290701  1.000000  0.411395  0.430208   
...        ...       ...       ...       ...       ...       ...       ...   
2315  0.063449  0.038069  0.133092  0.092673  0.121231  0.196724  0.191533   
2316  0.140859  0.190160  0.188025  0.295748  0.246709  0.218368  0.236228   
2317  0.215965  0.221648  0.379316  0.238653  0.289572  0.385054  0.381246   
2318  0.120468  0.112938  0.100504  0.178704  0.131871  0.121046  0.101015   
2319  0.164845  0.181330  0.251478  0.190611  0.236225  0.176678  0.230378   

          7         8         9     ...      2310      2311      2312  \
0     0.313786  0.296174  0.255994  ...  0.280447  0.280447  0.232048   
1     0.294174  0.310983  0.248465  ...  0.258874  0.258874  0.215473   
2     0.349047  0.291804  0.186650  ...  0.560502  0.560502  0.305534   
3     0.226770  0.378517  0.288675  ...  0.329861  0.329861  0.232004   
4     0.333080  0.337965  0.263743  ...  0.437950  0.437950  0.316692   
...        ...       ...       ...  ...       ...       ...       ...   
2315  0.104524  0.169128  0.103188  ...  0.388047  0.388047  0.441695   
2316  0.116024  0.187735  0.343622  ...  0.239300  0.239300  0.140083   
2317  0.160500  0.302984  0.184856  ...  0.634479  0.634479  0.542589   
2318  0.159473  0.133798  0.224490  ...  0.328914  0.328914  0.539118   
2319  0.206904  0.244115  0.163833  ...  0.448078  0.448078  0.480881   

          2313      2314      2315      2316      2317      2318      2319  
0     0.236940  0.166667  0.063449  0.140859  0.215965  0.120468  0.164845  
1     0.266557  0.200000  0.038069  0.190160  0.221648  0.112938  0.181330  
2     0.329456  0.174801  0.133092  0.188025  0.379316  0.100504  0.251478  
3     0.369504  0.213003  0.092673  0.295748  0.238653  0.178704  0.190611  
4     0.369404  0.212298  0.121231  0.246709  0.289572  0.131871  0.236225  
...        ...       ...       ...       ...       ...       ...       ...  
2315  0.383356  0.437798  1.000000  0.096523  0.467335  0.395554  0.401635  
2316  0.287860  0.126773  0.096523  1.000000  0.187327  0.209991  0.125388  
2317  0.555471  0.579695  0.467335  0.187327  1.000000  0.446736  0.494663  
2318  0.548572  0.609868  0.395554  0.209991  0.446736  1.000000  0.461710  
2319  0.488229  0.576958  0.401635  0.125388  0.494663  0.461710  1.000000  

[2320 rows x 2320 columns]

We can see that the cosine similarity scores on the diagonals are 1, this means that product 0 is 100% similar to itself.  We can see that product 0 is more similar to product 4 (cosine similarity score: 0.3892) than to product 3 (cosine similarity score: 0.2231). 

In [15]:
# Inputting into the model
# Select an arbitrary product asin to test
magazine_user_likes = 'B00005N7NQ'

def get_index_from_asin(asin):
    # Get the index of the product from it's asin
    return metaDF[metaDF.asin == asin].index.values[0]

magazine_index = get_index_from_asin(magazine_user_likes)

In [16]:
# Generate similar magazines matrix
similar_magazines = list(enumerate(cosine_sim[magazine_index]))

In [17]:
#sort similar magazines list in descending order
sorted_similar_magazines = sorted(similar_magazines, key=lambda x:x[1], reverse=True)
pd.DataFrame(sorted_similar_magazines)

0         1
0        0  1.000000
1      543  0.527046
2      334  0.523249
3     1578  0.507833
4      376  0.487518
...    ...       ...
2315  2212  0.000000
2316  2213  0.000000
2317  2215  0.000000
2318  2235  0.000000
2319  2253  0.000000

[2320 rows x 2 columns]

In [18]:
# print categories associated with the chosen input magazine
metaDF[metaDF['asin'] == 'B00005N7NQ']['category'][0]

"['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'Economics & Economic Theory']"

In [19]:
# Print top 15 most similar magazines

pd.set_option('display.max_colwidth', None)


def get_magazine_from_index(index):
    return metaDF.iloc[index].loc['asin'], metaDF.iloc[index].loc['category']

i = 0
for magazine in sorted_similar_magazines:
    print(get_magazine_from_index(magazine[0]))
    i += 1
    if i > 15:
        break

('B00005N7NQ', "['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'Economics & Economic Theory']")
('B00006L19P', "['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'History']")
('B00006KBGM', "['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'Economics & Economic Theory']")
('B00AA0B180', "['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'Economics & Economic Theory']")
('B00006K452', "['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'Economics & Economic Theory']")
('B01E1SNEHM', "['Magazine Subscriptions', 'Professional & Educational Journals', 'Professional & Trade', 'Humanities & Social Sciences', 'Economics & Economic Theor

We can see that the recommended magazines are all categorized as 'Professional & Educational Journals' and 'Professional & Trade.'  This is the same as the input magazine we fed into the model.  We can see that these recommendations make sense.